src/utils/charts.js

```diff
@@ -0,0 +1,241 @@
+const _ = require('lodash');
+const {
+  LEARNING_ANALYTICS_USER_ID,
+  MIN_PERCENTAGE_TO_SHOW_VERB,
+  OTHER_VERB,
+  LATE_NIGHT,
+  EARLY_MORNING,
+  MORNING,
+  AFTERNOON,
+  EVENING,
+  NIGHT,
+} = require('../config/constants');
+
+// Takes array of action objects and returns an object with {key: value} pairs of {date: #-of-actions}
+export const getActionsByDay = (actions) => {
+  const actionsByDay = {};
+  actions.forEach((action) => {
+    const actionDate = new Date(action.createdAt.slice(0, 10));
+    if (!actionsByDay[actionDate]) {
+      actionsByDay[actionDate] = 1;
+    } else {
+      actionsByDay[actionDate] += 1;
+    }
+  });
+  return actionsByDay;
+};
+
+// Takes object with {key: value} pairs of {date: #-of-actions} and returns a date-sorted array in Recharts.js format
+export const formatActionsByDay = (actionsByDayObject) => {
+  const actionsByDayArray = Object.entries(actionsByDayObject);
+  const sortedActionsByDay = actionsByDayArray.sort(
+    (entryA, entryB) => Date.parse(entryA[0]) - Date.parse(entryB[0]),
+  );
+  return sortedActionsByDay.map((entry) => {
+    const entryDate = new Date(entry[0]);
+    return {
+      date: `${entryDate.getDate()}-${
+        entryDate.getMonth() + 1
+      }-${entryDate.getFullYear()}`,
+      count: entry[1],
+    };
+  });
+};
+
+// given an actionsByDay object, finds max value to set on the yAxis in the graph in ActionsByDayChart.js
+export const findYAxisMax = (actionsByDay) => {
+  const arrayOfActionsCount = Object.values(actionsByDay);
+  if (!arrayOfActionsCount.length) {
+    return null;
+  }
+  const maxActionsCount = arrayOfActionsCount.reduce((a, b) => Math.max(a, b));
+  let yAxisMax;
+  // if maxActionsCount <= 100, round up yAxisMax to the nearest ten; else, to the nearest hundred
+  if (maxActionsCount <= 100) {
+    yAxisMax = Math.ceil(maxActionsCount / 10) * 10;
+  } else {
+    yAxisMax = Math.ceil(maxActionsCount / 100) * 100;
+  }
+  return yAxisMax;
+};
+
+// helper function used in getActionsByTimeOfDay
+// todo: update this function to retrieve hour of day using JS Date objects/moment
+const getActionHourOfDay = (action) => {
+  // expects action to be an object with a createdAt key
+  // createdAt should have the format "2020-12-31T23:59:59.999Z"
+  return action.createdAt.slice(11, 13);
+};
+
+// Takes array of action objects and returns an object with {key: value} pairs of {hourOfDay: #-of-actions}
+export const getActionsByTimeOfDay = (actions) => {
+  const actionsByTimeOfDay = {
+    [LATE_NIGHT]: 0,
+    [EARLY_MORNING]: 0,
+    [MORNING]: 0,
+    [AFTERNOON]: 0,
+    [EVENING]: 0,
+    [NIGHT]: 0,
+  };
+  actions.forEach((action) => {
+    const actionHourOfDay = getActionHourOfDay(action);
+    if (actionHourOfDay >= 0 && actionHourOfDay < 4) {
```

Comment:
Minor comment: maybe defining the limits for each period in constants would help in case we would like to have less categories, etc..

Reply:
good idea, but I think tricky in practice. if we have something like `START_OF_DAY = 0`, `END_OF_MORNING = 4`, etc, that might work, but then both this `if/else` chain and the `LATE_NIGHT`, `EARLY_MORNING`, etc, constants would have to be refactored/redefined, so it gets a bit circular.